StorageException: "Impossible store length" when storing 256 KB string using DynamoDbConnector

## Environment Details
* EclipseStore Version: 2.1.3
* JDK version: AdoptiumOpenJdk-21.0.4+7
* OS: Windows 11 Enterprise 24H2 26100.3775
* Used frameworks: AWS SDK 2.30.11

## Describe the bug
Exception is raised when storing relatively large objects, e.g. 256 KB strings using DynamoDbConnector.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>es-ddb-test</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>21</maven.compiler.source>
        <maven.compiler.target>21</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.eclipse.store</groupId>
            <artifactId>storage-embedded</artifactId>
            <version>2.1.3</version>
        </dependency>
        <dependency>
            <groupId>org.eclipse.store</groupId>
            <artifactId>afs-aws-dynamodb</artifactId>
            <version>2.1.3</version>
        </dependency>
        <dependency>
            <groupId>software.amazon.awssdk</groupId>
            <artifactId>dynamodb</artifactId>
            <version>2.30.11</version>
        </dependency>
    </dependencies>

</project>
```

```java
package org.example;

import org.eclipse.store.afs.aws.dynamodb.types.DynamoDbConnector;
import org.eclipse.store.afs.blobstore.types.BlobStoreFileSystem;
import org.eclipse.store.storage.embedded.types.EmbeddedStorage;
import org.eclipse.store.storage.types.Storage;
import org.eclipse.store.storage.types.StorageChannelCountProvider;
import org.eclipse.store.storage.types.StorageConfiguration;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;

import java.util.Arrays;
import java.util.Date;

public class Main {

  public static class DataRoot {
    private String content;

    public DataRoot() {
      super();
    }

    public String getContent() {
      return this.content;
    }

    public void setContent(final String content) {
      this.content = content;
    }

    @Override
    public String toString() {
      return "Root: " + this.content;
    }
  }

  public static void main(String[] args) {

    var fs = BlobStoreFileSystem.New(
            DynamoDbConnector.New(
                    DynamoDbClient.builder()
                            .credentialsProvider(
                                    DefaultCredentialsProvider.create()
                            )
                            .build()
            ));

    var fileProvider = Storage.FileProviderBuilder(fs)
            .setDirectory(fs.ensureDirectoryPath("es-db"))
            .createFileProvider();

    var storageConfig = StorageConfiguration.Builder()
            .setStorageFileProvider(fileProvider)
            .setChannelCountProvider(StorageChannelCountProvider.New(1));


    final DataRoot root = new DataRoot();
    var storage =  EmbeddedStorage
            .Foundation(storageConfig).start(root);

    var array = new byte[256 * 1024]; // 256 KB array
    Arrays.fill(array, (byte) 1);
    var content = new String(array);

    System.out.println("Root: " + root.getContent());

    root.setContent(content);

    storage.storeRoot();

    storage.shutdown();
  }
}
```

## To Reproduce
1. Configure the AWS credentials if needed
2. Run the code 3+ times.

Ignore the first time with ResoureNotFoundException already described in issue #399.
Second execution might complete normally when data is saved.
Third execution results in the following exception.

```
Exception in thread "main" org.eclipse.serializer.persistence.exceptions.PersistenceExceptionTransfer
	at org.eclipse.store.storage.embedded.types.EmbeddedStorageBinaryTarget$Default.write(EmbeddedStorageBinaryTarget.java:84)
	at org.eclipse.store.storage.embedded.types.EmbeddedStorageBinaryTarget$Default.write(EmbeddedStorageBinaryTarget.java:43)
	at org.eclipse.serializer.persistence.binary.types.BinaryStorer$Default.commit(BinaryStorer.java:580)
	at org.eclipse.serializer.persistence.types.PersistenceManager$Default.storeAll(PersistenceManager.java:320)
	at org.eclipse.store.storage.types.StorageConnection.storeAll(StorageConnection.java:411)
	at org.eclipse.store.storage.embedded.types.EmbeddedStorageManager$Default.storeRoot(EmbeddedStorageManager.java:194)
	at org.example.Main.main(Main.java:76)
Caused by: org.eclipse.store.storage.exceptions.StorageException: Problem in channel #0
	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.checkForProblems(StorageChannelTask.java:106)
	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.waitOnCompletion(StorageChannelTask.java:168)
	at org.eclipse.store.storage.types.StorageRequestAcceptor$Default.waitOnTask(StorageRequestAcceptor.java:159)
	at org.eclipse.store.storage.types.StorageRequestAcceptor$Default.storeData(StorageRequestAcceptor.java:177)
	at org.eclipse.store.storage.embedded.types.EmbeddedStorageBinaryTarget$Default.write(EmbeddedStorageBinaryTarget.java:80)
	... 6 more
Caused by: org.eclipse.store.storage.exceptions.StorageException: Impossible store length:
timestamp = 1745417881314000000
currentTotalLength = 2648
uncommittedDataLength = 524392
resulting length = 527040
dataBuffers: 
#0: limit = 64, position = 64, capacity = 4096;
#1: limit = 524328, position = 524328, capacity = 524328;
	at org.eclipse.store.storage.types.StorageFileManager$Default.throwImpossibleStoreLengthException(StorageFileManager.java:1793)
	at org.eclipse.store.storage.types.StorageFileManager$Default.storeChunks(StorageFileManager.java:644)
	at org.eclipse.store.storage.types.StorageChannel$Default.storeEntities(StorageChannel.java:592)
	at org.eclipse.store.storage.types.StorageRequestTaskStoreEntities$Default.internalProcessBy(StorageRequestTaskStoreEntities.java:73)
	at org.eclipse.store.storage.types.StorageRequestTaskStoreEntities$Default.internalProcessBy(StorageRequestTaskStoreEntities.java:40)
	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.processBy(StorageChannelTask.java:244)
	at org.eclipse.store.storage.types.StorageChannel$Default.work(StorageChannel.java:453)
	at org.eclipse.store.storage.types.StorageChannel$Default.run(StorageChannel.java:536)
	at java.base/java.lang.Thread.run(Thread.java:1583)

Process finished with exit code 1
```

## Expected behavior
Data should be stored and read correctly.

--- a/afs/aws/dynamodb/src/main/java/org/eclipse/store/afs/aws/dynamodb/types/DynamoDbConnector.java
+++ b/afs/aws/dynamodb/src/main/java/org/eclipse/store/afs/aws/dynamodb/types/DynamoDbConnector.java
@@ -50,8 +50,8 @@ import software.amazon.awssdk.services.dynamodb.model.QueryResponse;
 import software.amazon.awssdk.services.dynamodb.model.ResourceNotFoundException;
 import software.amazon.awssdk.services.dynamodb.model.ScalarAttributeType;
 import software.amazon.awssdk.services.dynamodb.model.ScanRequest;
+import software.amazon.awssdk.services.dynamodb.model.ScanResponse;
 import software.amazon.awssdk.services.dynamodb.paginators.QueryIterable;
-import software.amazon.awssdk.services.dynamodb.paginators.ScanIterable;
 import software.amazon.awssdk.services.dynamodb.model.Select;
 import software.amazon.awssdk.services.dynamodb.model.TableDescription;
 import software.amazon.awssdk.services.dynamodb.model.TableStatus;
@@ -327,14 +327,33 @@ public interface DynamoDbConnector extends BlobStoreConnector
 				 * the first page would silently hide files from {@link
 				 * org.eclipse.store.storage.types.StorageFileManager}'s startup
 				 * inventory and cause "Non-deleted non-empty data file not found".
+				 *
+				 * We drain the paginator into a list here, inside the try block, so
+				 * a {@link ResourceNotFoundException} thrown on any page is caught
+				 * consistently with {@link #blobs(BlobStorePath, boolean)} regardless
+				 * of whether the AWS SDK paginator fetches eagerly or lazily — the
+				 * earlier version that returned a lazy stream straight out of the
+				 * try/catch relied on the SDK's current (eager) page-fetch behavior
+				 * and would silently leak the exception if that ever changed.
 				 */
 				final Pattern pattern = Pattern.compile(childKeysRegexWithContainer(directory));
-				final ScanIterable pages = this.client.scanPaginator(request);
-				return pages.items().stream()
-					.map(item -> item.get(FIELD_KEY).s())
-					.filter(key -> pattern.matcher(key).matches())
-					.distinct()
-				;
+				final List<String> keys = new ArrayList<>();
+				for(final ScanResponse page : this.client.scanPaginator(request))
+				{
+					if(!page.hasItems())
+					{
+						continue;
+					}
+					for(final Map<String, AttributeValue> item : page.items())
+					{
+						final String key = item.get(FIELD_KEY).s();
+						if(pattern.matcher(key).matches())
+						{
+							keys.add(key);
+						}
+					}
+				}
+				return keys.stream().distinct();
 			}
 			catch(final ResourceNotFoundException e)
 			{

--- a/afs/aws/dynamodb/src/main/java/org/eclipse/store/afs/aws/dynamodb/types/DynamoDbConnector.java
+++ b/afs/aws/dynamodb/src/main/java/org/eclipse/store/afs/aws/dynamodb/types/DynamoDbConnector.java
@@ -17,7 +17,6 @@ package org.eclipse.store.afs.aws.dynamodb.types;
 import static org.eclipse.serializer.util.X.checkArrayRange;
 import static org.eclipse.serializer.util.X.notNull;
 
-import java.io.BufferedInputStream;
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.util.ArrayList;
@@ -30,7 +29,6 @@ import java.util.stream.Stream;
 import org.eclipse.serializer.concurrency.XThreads;
 import org.eclipse.serializer.exceptions.IORuntimeException;
 import org.eclipse.serializer.io.ByteBufferInputStream;
-import org.eclipse.serializer.io.LimitedInputStream;
 import org.eclipse.store.afs.blobstore.types.BlobStoreConnector;
 import org.eclipse.store.afs.blobstore.types.BlobStorePath;
 
@@ -52,7 +50,8 @@ import software.amazon.awssdk.services.dynamodb.model.QueryResponse;
 import software.amazon.awssdk.services.dynamodb.model.ResourceNotFoundException;
 import software.amazon.awssdk.services.dynamodb.model.ScalarAttributeType;
 import software.amazon.awssdk.services.dynamodb.model.ScanRequest;
-import software.amazon.awssdk.services.dynamodb.model.ScanResponse;
+import software.amazon.awssdk.services.dynamodb.paginators.QueryIterable;
+import software.amazon.awssdk.services.dynamodb.paginators.ScanIterable;
 import software.amazon.awssdk.services.dynamodb.model.Select;
 import software.amazon.awssdk.services.dynamodb.model.TableDescription;
 import software.amazon.awssdk.services.dynamodb.model.TableStatus;
@@ -229,16 +228,27 @@ public interface DynamoDbConnector extends BlobStoreConnector
 			
 			try
 			{
-				final QueryResponse response = this.client.query(builder.build());
-				if(!response.hasItems())
+				/*
+				 * DynamoDB queries return paginated results capped at 1 MB per page.
+				 * A multi-blob file's seq rows (each up to MAX_BLOB_SIZE = 400 000 B)
+				 * easily exceed one page, so we MUST consume the full paginator,
+				 * otherwise file size / existence / read-back will be silently
+				 * truncated to the first page.
+				 */
+				final QueryIterable pages = this.client.queryPaginator(builder.build());
+				final List<Map<String, AttributeValue>> items = new ArrayList<>();
+				for(final QueryResponse page : pages)
+				{
+					if(page.hasItems())
+					{
+						items.addAll(page.items());
+					}
+				}
+				if(items.isEmpty())
 				{
 					return Stream.empty();
 				}
-								
-				return response.items()
-					.stream()
-					.sorted(this.blobComparator())
-				;
+				return items.stream().sorted(this.blobComparator());
 			}
 			catch(final ResourceNotFoundException e)
 			{
@@ -307,19 +317,20 @@ public interface DynamoDbConnector extends BlobStoreConnector
 				.expressionAttributeValues(expressionValues)
 				.build()
 			;
-			
-			final ScanResponse response = this.client.scan(request);
+
 			try
 			{
-				if(!response.hasItems())
-				{
-					return Stream.empty();
-				}
-
+				/*
+				 * DynamoDB scans page at ~1 MB per response. A directory containing
+				 * multiple multi-blob files (a normal Eclipse Store channel does)
+				 * exceeds that, so we MUST iterate the full paginator. Returning only
+				 * the first page would silently hide files from {@link
+				 * org.eclipse.store.storage.types.StorageFileManager}'s startup
+				 * inventory and cause "Non-deleted non-empty data file not found".
+				 */
 				final Pattern pattern = Pattern.compile(childKeysRegexWithContainer(directory));
-				
-				return response.items()
-					.stream()
+				final ScanIterable pages = this.client.scanPaginator(request);
+				return pages.items().stream()
 					.map(item -> item.get(FIELD_KEY).s())
 					.filter(key -> pattern.matcher(key).matches())
 					.distinct()
@@ -446,23 +457,41 @@ public interface DynamoDbConnector extends BlobStoreConnector
 					MAX_BLOB_SIZE
 				);
 
-				try(LimitedInputStream limitedInputStream = LimitedInputStream.New(
-					new BufferedInputStream(buffersInputStream),
-					currentBatchSize)
-				)
+				try
 				{
+					/*
+					 * Read directly from the (in-memory) ByteBufferInputStream into the
+					 * batch array. Earlier versions wrapped this in BufferedInputStream,
+					 * but BufferedInputStream's 8 KiB internal pre-fetch buffer caused
+					 * silent data loss at blob boundaries: bytes pulled from the source
+					 * into the BIS buffer that were never returned to the caller were
+					 * discarded when the BIS was closed (try-with-resources, recreated
+					 * per batch). The next batch then started reading from a source
+					 * position that was up to 8 KiB past where it should have been.
+					 *
+					 * The offset into batch[] must also advance with each read; an
+					 * InputStream.read may legitimately return fewer bytes than
+					 * requested, and a fixed offset of 0 would overwrite earlier bytes.
+					 */
 					final byte[] batch = new byte[checkArrayRange(currentBatchSize)];
-						  int    remaining = batch.length;
-				          int    read;
-					while(remaining > 0 &&
-						(read = limitedInputStream.read(
+					      int    off   = 0;
+					      int    read;
+					while(off < batch.length &&
+						(read = buffersInputStream.read(
 							batch,
-							0,
-							Math.min(batch.length, remaining))
+							off,
+							batch.length - off)
 						) != -1
 					)
 					{
-						remaining -= read;
+						off += read;
+					}
+					if(off < batch.length)
+					{
+						throw new IORuntimeException(new IOException(
+							"Source stream exhausted: expected "
+							+ batch.length + " bytes, got " + off
+						));
 					}
 
 					final Map<String, AttributeValue> item = new HashMap<>();

